zap/src/parser/mod.rs

```diff
@@ -1,178 +1,102 @@
-use std::{collections::HashSet, path::PathBuf};
+use std::collections::HashSet;
 
-use lalrpop_util::lalrpop_mod;
+use codespan_reporting::diagnostic::{Diagnostic, Label, Severity};
+use lalrpop_util::{lalrpop_mod, ParseError};
 
-use crate::{
-	util::{NumTy, Range},
-	Error,
-};
+use crate::config::{Config, EvType};
 
-mod working_ast;
+use self::syntax_tree::{Spanned, SyntaxEvDecl};
 
-lalrpop_mod!(pub grammar);
-
-#[derive(Debug)]
-pub struct File {
-	pub ty_decls: Vec<TyDecl>,
-	pub ev_decls: Vec<EvDecl>,
-
-	pub server_output: PathBuf,
-	pub client_output: PathBuf,
-
-	pub casing: Casing,
-	pub typescript: bool,
-	pub write_checks: bool,
-}
-
-impl File {
-	pub fn event_id_ty(&self) -> NumTy {
-		NumTy::from_f64(0.0, self.ev_decls.len() as f64)
-	}
-}
-
-#[derive(Debug, Clone, Copy)]
-pub enum Casing {
-	Pascal,
-	Camel,
-	Snake,
-}
-
-#[derive(Debug, Clone)]
-pub struct EvDecl {
-	pub name: String,
-	pub from: EvSource,
-	pub evty: EvType,
-	pub call: EvCall,
-	pub data: Ty,
-}
-
-#[derive(Debug, Clone, Copy, PartialEq, Eq)]
-pub enum EvSource {
-	Server,
-	Client,
-}
-
-#[derive(Debug, Clone, Copy, PartialEq, Eq)]
-pub enum EvType {
-	Reliable,
-	Unreliable,
-}
-
-#[derive(Debug, Clone, Copy)]
-pub enum EvCall {
-	SingleSync,
-	SingleAsync,
-	ManySync,
-	ManyAsync,
-}
-
-#[derive(Debug, Clone)]
-pub struct TyDecl {
-	pub name: String,
-	pub ty: Ty,
-}
-
-#[derive(Debug, Clone)]
-pub enum Ty {
-	Bool,
-
-	F32(Range<f32>),
-	F64(Range<f64>),
-
-	I8(Range<i8>),
-	I16(Range<i16>),
-	I32(Range<i32>),
-
-	U8(Range<u8>),
-	U16(Range<u16>),
-	U32(Range<u32>),
-
-	Str { len: Range<u16> },
-	Arr { len: Range<u16>, ty: Box<Ty> },
-	Map { key: Box<Ty>, val: Box<Ty> },
-
-	Struct { fields: Vec<(String, Ty)> },
-	Enum { variants: Vec<String> },
+mod convert;
+mod syntax_tree;
 
-	Instance(bool, Option<String>),
-	Vector3,
-
-	Ref(String),
-
-	Optional(Box<Ty>),
-}
-
-impl Ty {
-	pub fn exact_size(&self) -> Option<usize> {
-		match self {
-			Ty::Bool => Some(1),
-
-			Ty::F32(_) => Some(4),
-			Ty::F64(_) => Some(8),
-
-			Ty::I8(_) => Some(1),
-			Ty::I16(_) => Some(2),
-			Ty::I32(_) => Some(4),
-
-			Ty::U8(_) => Some(1),
-			Ty::U16(_) => Some(2),
-			Ty::U32(_) => Some(4),
-
-			Ty::Str { len } => len.exact().map(|len| len as usize),
+lalrpop_mod!(pub grammar);
 
-			Ty::Arr { len, ty } => {
-				if let Some(len) = len.exact() {
-					ty.exact_size().map(|ty_size| len as usize * ty_size)
-				} else {
-					None
+pub fn parse(input: &str) -> (Option<Config<'_>>, Vec<Diagnostic<()>>) {
+	let parse_result = grammar::ConfigParser::new().parse(input);
+
+	if let Ok(syntax_config) = parse_result {
+		let evdecls: Vec<SyntaxEvDecl<'_>> = syntax_config
+			.decls
+			.iter()
+			.filter_map(|decl| match decl {
+				syntax_tree::SyntaxDecl::Ev(evdecl) => Some(evdecl.clone()),
+				_ => None,
+			})
+			.collect();
+
+		let (config, mut diags) = convert::convert(syntax_config);
```

Comment:
What's `evdecls` used for if `convert::convert` also returns the `evdecls`?

Reply:
evdecls is used to keep the syntax tree for evdecls - otherwise we would be unable to error if we find that an evdecl is above the data limit.